### PR TITLE
Extend trivial count optimization.

### LINKED ISSUE #11092: when set max_parallel_replicas=n, count query's result change to n*correct result.
(you don't have to strictly follow this form)

**Describe the bug**
I have as cluster,one distribute node,two local node to consist  a shard.
when set max_parallel_replicas=n,count query'sresult change to n*correct result.
**How to reproduce**
use a replicated table,and one shard has tow or more replicas,then set set max_parallel_replicas=n.
* Which ClickHouse server version to use
20.1.6.30

* Queries to run that lead to unexpected result
>**select count()from default.iplog;**
>this count result is 387000.
>**then set max_parallel_replicas=2;**
>execute the same query.
>**select count()from default.iplog;**
>this count result is 774000.
**Expected behavior**
why is the result not 387000 but 774000 when  i set max_parallel_replicas=2?Obveriously,the result is the sum of two replicas's count result:387000 +387000 =774000 .
can we select one of the max result or most correct result of replicas?

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now trivial count optimization becomes slightly non-trivial. Predicates that contain exact partition expr can be optimized too. This also fixes https://github.com/ClickHouse/ClickHouse/issues/11092 which returns wrong count when `max_parallel_replicas > 1`.

Detailed description / Documentation draft:

None.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
> test_rename_column

This test was broken and is already fixed in master

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -3554,6 +3554,36 @@ std::optional<UInt64> StorageReplicatedMergeTree::totalRows() const
     return res;
 }
 
+std::optional<UInt64> StorageReplicatedMergeTree::totalRowsByPartitionPredicate(const SelectQueryInfo & query_info, const Context & context) const
```
Comment (alexey-milovidov) [inline src/Storages/StorageReplicatedMergeTree.cpp:3557]: It's 100% copy-paste, need to move common code to `MergeTreeData`.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: src/Storages/MergeTree/KeyCondition.h [hunk 2]
```diff
@@ -413,6 +414,9 @@ class KeyCondition
     ColumnIndices key_columns;
     ExpressionActionsPtr key_expr;
     PreparedSets prepared_sets;
+
+    // If true, do not take any function monotonicity into account
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/KeyCondition.h:418]: I think it's not only about the monotonicity...

Will it work correctly for expressions like:
```
SELECT count() FROM table
WHERE partition = 123 AND sipHash64(toString(partition)) % 2 = 1
```
?

(When there are multiple conditions. All of them depend only on columns in partition key. But there are some non-analyzable conditions.)

Comment (alexey-milovidov) [reply src/Storages/MergeTree/KeyCondition.h:418]: There are no tests for this case.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -3554,6 +3554,36 @@ std::optional<UInt64> StorageReplicatedMergeTree::totalRows() const
...
+{
+    auto metadata_snapshot = getInMemoryMetadataPtr();
+    const auto & partition_key = metadata_snapshot->getPartitionKey();
+    Names partition_key_columns = partition_key.column_names;
+    KeyCondition key_condition(query_info, context, partition_key_columns, partition_key.expression, true /* exact_match */);
+    if (key_condition.alwaysUnknownOrTrue())
+        return {};
+    std::unordered_map<String, bool> partition_filter_map;
+    size_t res = 0;
+    foreachCommittedParts([&](auto & part)
```
Comment (alexey-milovidov) [inline src/Storages/StorageReplicatedMergeTree.cpp:3567]: We can analyze partitions first by simply evaluating the condition (with VirtualColumnUtils) on partition expression - to count or filter out whole partitions before filtering parts.

Comment (alexey-milovidov) [reply src/Storages/StorageReplicatedMergeTree.cpp:3567]: I was wrong.

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 1 above)
Comment (amosbird) [reply src/Storages/StorageReplicatedMergeTree.cpp:3566]: I didn't find a good way to abstract, and is worry about introducing non-trivial template code.

File: src/Storages/MergeTree/KeyCondition.h (hunk 2 above)
Comment (amosbird) [reply src/Storages/MergeTree/KeyCondition.h]: Actually we can apply any function transformations to single point range.

**Pull request merged**